opentelemetry-sdk/src/resource/builder.rs

```diff
@@ -0,0 +1,102 @@
+use std::time::Duration;
+
+use opentelemetry::KeyValue;
+
+use super::{Resource, ResourceDetector};
+
+/// Builder to allow easy composition of a Resource
+#[derive(Debug)]
+pub struct ResourceBuilder {
+    resource: Resource,
+}
+
+impl Default for ResourceBuilder {
+    /// Create ResourceBuilder with [Resource::default()].
+    fn default() -> Self {
+        ResourceBuilder {
+            resource: Resource::default(),
+        }
+    }
+}
+
+impl ResourceBuilder {
+    /// Create ResourceBuilder with [Resource::empty()].
+    pub fn new() -> Self {
+        ResourceBuilder {
+            resource: Resource::empty(),
+        }
+    }
+
+    /// Add a single [ResourceDetector] to your resource.
+    pub fn with_detector(self, detector: Box<dyn ResourceDetector>) -> Self {
+        self.with_detectors(vec![detector])
+    }
+
+    /// Add multiple [ResourceDetector] to your resource.
+    pub fn with_detectors(mut self, detectors: Vec<Box<dyn ResourceDetector>>) -> Self {
+        self.resource = self
+            .resource
+            .merge(&Resource::from_detectors(Duration::from_secs(0), detectors));
+        self
+    }
+
+    /// Add a [KeyValue] to the resource.
+    pub fn with_key_value(self, kv: KeyValue) -> Self {
+        self.with_key_values(vec![kv])
+    }
+
+    /// Add multiple [KeyValue]s to the resource.
+    pub fn with_key_values<T: IntoIterator<Item = KeyValue>>(mut self, kvs: T) -> Self {
+        self.resource = self.resource.merge(&Resource::new(kvs));
+        self
+    }
+
```

Comment:
should there also be a method `with_schema_url` to add schema url ?

Reply:
How would we like the behavior of the `with_schema_url` to act? Do we want to follow the rules that `from_schema_url` has? Or override the current one with the specified one, or maybe use TypeState pattern to only allow the schema_url to be set once?

Reply:
this is my initial thought
```rust
/// This will merge the provided `schema_url` with the current state of the Resource being built. It
/// will use the following rules to determine which `schema_url` should be used.
///
/// ### [Schema url]
/// If both of the resources are not empty. Schema url is determined by the following rules, in order:
/// 1. If the provided resource has a schema url, it will be used.
/// 2. If the provided resource does not have a schema url, and the builders current resource has a schema url, it will be used.
/// 3. If both resources have a schema url and it's the same, it will be used.
/// 4. If both resources have a schema url and it's different, the schema url will be empty.
/// 5. If both resources do not have a schema url, the schema url will be empty.
///
/// [Schema url]: https://github.com/open-telemetry/opentelemetry-specification/blob/v1.9.0/specification/schemas/overview.md#schema-url
pub fn with_schema_url<S>(mut self, schema_url: S) -> Self
where
    S: Into<Cow<'static, str>>,
{
    self.resource = Resource::from_schema_url([], schema_url).merge(&self.resource);
    self
}
```

Reply:
added this, would love to hear feedback on the implementation!